zebra/zebra_gr.c

```diff
@@ -183,10 +315,412 @@ zebra_gr_client_reconnect(struct zserv *client)
+	default:
+		break;
+	}
+}
+
+/*
+ * Function to decode and call appropriate functions
+ * to handle client capabilities.
+ */
+void
+zread_client_capabilities(ZAPI_HANDLER_ARGS)
+{
+	struct zapi_cap api;
+	struct stream *s;
+
+	s = msg;
 
+	if (zapi_capabilities_decode(s, &api)) {
+		LOG_GR("%s: Error in reading capabilities for client %s",
+		       __func__,  zebra_route_string(client->proto));
+		return;
+	}
+
+	/* Call the capabilities handler */
+	zebra_client_capabilities_handler(client, &api);
+}
+
+
+
+/*******************************************
+ *		Stale route handling	   *
+ *******************************************/
+
+/*
+ * Delete all the stale routes that have not been refreshed
+ * post restart.
+ */
 static int
 zebra_gr_route_stale_delete_timer_expiry(struct thread *thread)
 {
+	struct client_gr_info *info;
+	int cnt = 0;
+	struct zserv *client;
+
+	info = THREAD_ARG(thread);
+	info->t_stale_removal = NULL;
+	client = (struct zserv *)info->stale_client_ptr;
+
+	/* Set the flag to indicate all stale route deletion */
+	if (thread->u.val == 1)
+		info->delete = true;
+
+	cnt = zebra_gr_delete_stale_routes(info);
+
+	/* Retsart the timer */
+	if (cnt > 0) {
+		LOG_GR("%s: Client %s processed %d routes. Start timer again",
+		       __func__, zebra_route_string(client->proto), cnt);
+
+		thread_add_timer(zrouter.master,
+				 zebra_gr_route_stale_delete_timer_expiry,
+				 info, ZEBRA_DEFAULT_STALE_UPDATE_DELAY,
+				 &info->t_stale_removal);
+	} else {
+		/* No routes to delete for the VRF */
+		LOG_GR("%s: Client %s all starle routes processed",
+		       __func__, zebra_route_string(client->proto));
+
+		if (info->current_prefix != NULL)
+			XFREE(MTYPE_TMP, info->current_prefix);
+		info->current_prefix = NULL;
+		info->current_afi = 0;
+		zebra_gr_delete_stale_client(info);
+	}
 	return 0;
 }
 
+
+
+/*
+ * Function to process to check if route entry is stale
+ * or has been updated.
+ */
+static void
+zebra_gr_process_route_entry(struct zserv *client,
+			      struct route_node *rn,
+			      struct route_entry *re)
+{
+	char buf[PREFIX2STR_BUFFER];
+
+	if ((client == NULL) || (rn == NULL) || (re == NULL))
+		return;
+
+	/* If the route is not refreshed after restart, delete the entry */
+	if (re->uptime < client->restart_time) {
+		if (IS_ZEBRA_DEBUG_RIB) {
+			prefix2str(&rn->p, buf, sizeof(buf));
+			zlog_debug("%s: Client %s stale route %s is deleted",
+				     __func__,
+				     zebra_route_string(client->proto), buf);
+		}
+		rib_delnode(rn, re);
+	}
+}
+
+/*
+ * This function walks through the route table for all vrf and deletes
+ * the stale routes for the restarted client specified by the protocol
+ * type
+ */
+static int
+zebra_gr_delete_stale_route(struct client_gr_info *info,
+			    struct zebra_vrf *zvrf)
+{
+	struct route_node *rn, *curr;
+	struct route_entry *re;
+	struct route_entry *next;
+	struct route_table *table;
+	int n = 0;
+	struct prefix *p;
+	afi_t afi, curr_afi;
+	unsigned char proto;
```

Comment:
uint8_t and uint16_t also please go through entire commit set and make sure we are consistent.

Reply:
Tanks. I will fix it.